db/migrate/20210406161242_migrate_enterprise_fee_tax_amounts.rb

```diff
@@ -0,0 +1,80 @@
+class MigrateEnterpriseFeeTaxAmounts < ActiveRecord::Migration[5.0]
+  class Spree::Adjustment < ActiveRecord::Base
+    belongs_to :originator, -> { with_deleted }, polymorphic: true
+    belongs_to :adjustable, polymorphic: true
+    belongs_to :order, class_name: "Spree::Order"
+    belongs_to :tax_category, class_name: 'Spree::TaxCategory'
+    has_many :adjustments, as: :adjustable, dependent: :destroy
+
+    scope :enterprise_fee, -> { where(originator_type: 'EnterpriseFee') }
+  end
+  class EnterpriseFee < ActiveRecord::Base
+    belongs_to :tax_category, class_name: 'Spree::TaxCategory', foreign_key: 'tax_category_id'
+  end
+  class Spree::LineItem < ActiveRecord::Base
+    belongs_to :variant, class_name: "Spree::Variant"
+    has_one :product, through: :variant
+  end
+  class Spree::Variant < ActiveRecord::Base
+    belongs_to :product, class_name: 'Spree::Product'
+    has_many :line_items, inverse_of: :variant
+  end
+  class Spree::Product < ActiveRecord::Base
+    belongs_to :tax_category, class_name: 'Spree::TaxCategory'
+    has_many :variants, class_name: 'Spree::Variant'
+  end
+
+  def up
+    migrate_enterprise_fee_taxes!
+  end
+
+  def migrate_enterprise_fee_taxes!
+    Spree::Adjustment.enterprise_fee.where('included_tax <> 0').includes(:originator).find_each do |fee|
```

Comment:
How long does this migration run for with production data?

Reply:
I haven't checked recently, I'll rebase and try it now :+1:

Reply:
For instances that use **additional tax**, there's not much that changes. This is the migration with CA prod data:
```
== 20210406161242 MigrateEnterpriseFeeTaxAmounts: migrating ===================
== 20210406161242 MigrateEnterpriseFeeTaxAmounts: migrated (2.3167s) ==========
```
This is because there aren't really any enterprise fee adjustments that have a non-zero value for `included_tax` :point_up: 

With French data it's very different. The total number of enterprise fee adjustments with non-zero included tax amounts is: **152,981.** 
```
== 20210406161242 MigrateEnterpriseFeeTaxAmounts: migrating ===================
== 20210406161242 MigrateEnterpriseFeeTaxAmounts: migrated (1668.5366s) =======
```
Total time: 1668 seconds (~27 minutes). :grimacing: 

I previously looked at doing this with raw SQL, but I think there's too much complexity...

Reply:
It's marginally when eager-loading both the originator and the adjustable:
```
== 20210406161242 MigrateEnterpriseFeeTaxAmounts: migrating ===================
== 20210406161242 MigrateEnterpriseFeeTaxAmounts: migrated (1405.0915s) =======
```
~23 minutes.

Reply:
Just wondering if we need to deploy this with maintenance. But I don't think that's necessary because this wouldn't cause any pages to raise errors, right?

Reply:
> deploy this with maintenance

:100:

Definitely! Active users could be changing objects that this migration is trying to alter. So for example new enterprise fee adjustments could be created using the "old" code which would then not be updated correctly by the migration. A 23 minute time window leaves plenty of room for creating lots of little data inconsistencies!

Reply:
Update: I just deployed this on `fr-staging` which currently has a very recent production dataset and it only took **~17 minutes**, which is a lot better. Maybe the ofn-install task that clears old data before migrating made a difference?